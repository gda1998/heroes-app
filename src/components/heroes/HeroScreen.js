import { useMemo } from 'react';
import { Navigate } from 'react-router-dom';
import { useNavigate, useParams } from 'react-router';
import { getHeroById } from '../../selectors/getHeroById';
import { heroeImages } from '../../helpers/heroesImages';

export const HeroScreen = () => {
    const { heroId } = useParams();
    const navigate = useNavigate();
    const hero = useMemo(() => getHeroById(heroId), [ heroId ]);
    const handleReturn = () => navigate(-1) === undefined && navigate('/');

    if(!hero) return <Navigate to="/" />

    const {
        superhero,
        publisher,
        alter_ego,
        first_appearance,
        characters
    } = hero;

    return (
        <div className="d-flex align-items-center animate__animated animate__fadeInLeft">
            {/* {
                console.log(heroeImages(heroId))
            } */}
            <div className="flex-shrink-0">
                <img loading="lazy" src={ heroeImages(heroId) } className="img-thumbnail" alt={superhero} />
            </div>
            <div className="flex-grow-1 ms-3">
                <h3>{ superhero }</h3>
                <div className="d-flex align-items-start flex-column bd-highlight mb-3" >
                    <div className="p-2 bd-highlight"><b>Alter Ego:</b> { alter_ego }</div>
                    <div className="p-2 bd-highlight"><b>Publisher:</b> { publisher }</div>
                    <div className="p-2 bd-highlight"><b>First appearance:</b> { first_appearance }</div>
                </div>
                <h5>Characters</h5>
                <p>{ characters }</p>
                <button className="btn btn-outline-info" onClick={ handleReturn }>Return</button>
            </div>
        </div>
    );
}
