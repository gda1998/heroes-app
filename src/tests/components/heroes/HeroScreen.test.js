import { mount } from 'enzyme';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { HeroScreen } from '../../../components/heroes/HeroScreen';

describe('Pruebas en <HeroScreen />', () => {
    beforeEach( () => jest.clearAllMocks() );

    const getWrapper = (heroId) => mount(
        <MemoryRouter initialEntries={[ `/hero/${heroId}` ]}>
            <Routes>
                <Route path="/hero/:heroId" element={ <HeroScreen /> } />
                <Route path="/" element={ <h1>Not Found</h1>} />
            </Routes>
        </MemoryRouter>
    );

    test('Debe de redireccionar a la pantalla de Marvel si no existe el heroe a buscar', () => {
        const wrapper = getWrapper('super-hombre');
        const existsMarvelTitle = wrapper.find('h1').exists();
        expect(existsMarvelTitle).toBe(true);
    });

    test('Debe de mostrar a Spiderman cuando el parametro exista', () => {
        const wrapper = getWrapper('marvel-spider');
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('h3').exists()).toBe(true);
        expect( wrapper.find('h3').text().trim() ).toBe('Spider Man');
    });
});
