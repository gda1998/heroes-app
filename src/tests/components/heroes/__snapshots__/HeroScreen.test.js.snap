// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HeroScreen /> Debe de mostrar a Spiderman cuando el parametro exista 1`] = `
<div
  className="d-flex align-items-center animate__animated animate__fadeInLeft"
>
  <div
    className="flex-shrink-0"
  >
    <img
      alt="Spider Man"
      className="img-thumbnail"
      loading="lazy"
    />
  </div>
  <div
    className="flex-grow-1 ms-3"
  >
    <h3>
      Spider Man
    </h3>
    <div
      className="d-flex align-items-start flex-column bd-highlight mb-3"
    >
      <div
        className="p-2 bd-highlight"
      >
        <b>
          Alter Ego:
        </b>
         
        Peter Parker
      </div>
      <div
        className="p-2 bd-highlight"
      >
        <b>
          Publisher:
        </b>
         
        Marvel Comics
      </div>
      <div
        className="p-2 bd-highlight"
      >
        <b>
          First appearance:
        </b>
         
        Amazing Fantasy #15
      </div>
    </div>
    <h5>
      Characters
    </h5>
    <p>
      Peter Parker
    </p>
    <button
      className="btn btn-outline-info"
      onClick={[Function]}
    >
      Return
    </button>
  </div>
</div>
`;
